Allow FieldTracker on Django user model (and other children of abstract base classes)

## Problem

Here is a branch with failing tests that I think should work (and that I think worked in earlier versions?): https://github.com/jazzband/django-model-utils/compare/master...jcushman:abstract-test-failure

This seems to be related to the stuff that @lucaswiman added in #317. @lucaswiman, I'm hoping based on that work you might have some clever idea for how to fix this. :)

Here's the situation: you have an abstract base class that defines a static attribute like `is_active = True`, and a concrete model inheriting from that class that defines a field like `is_active = models.BooleanField(default=True)`. The model then throws an error on `save()`:

```
from django.contrib.auth.models import AbstractUser

class MyUser(AbstractUser):
    tracker = FieldTracker()

MyUser().save()
```

Result:

```
    def __get__(self, instance, owner):
        if instance is None:
            return self
        was_deferred = self.field_name in instance.get_deferred_fields()
>       value = self.descriptor.__get__(instance, owner)
E       AttributeError: 'bool' object has no attribute '__get__'

model_utils/tracker.py:43: AttributeError
```

It would be great for this to work, because tracking changes on the Django user model is handy.

Debugging that error, I found the problem boils down to this:

```
class AbstractUser(models.Model):
    is_active = True

    class Meta:
        abstract = True

class MyUser(AbstractUser):
    is_active = models.BooleanField(default=True)
    tracker = FieldTracker()

MyUser().save()
```

The reason that fails is https://github.com/jazzband/django-model-utils/blob/master/model_utils/tracker.py#L218 :

```
descriptor = getattr(sender, field_name)
...
setattr(sender, field_name, wrapped_descriptor)
```

... which boils down to setting `MyUser.is_active = DescriptorWrapper(MyUser.is_active)`. And _that_ doesn't work because you expect `MyUser.is_active` to start with a value of `DeferredAttribute('is_active')`, but it actually returns `True`. For reasons I don't understand, when you override a static attribute on an abstract class, you get the base class's value back instead of the subclass's.

I tried tweaking tracker.py with variations on `descriptor = getattr(sender, field_name) if field_name in sender.__dict__ else DeferredAttribute(field_name)`, but that broke foreign key fields and feels pretty janky anyway.

Any ideas on how to get this working?

Thanks!

## Environment

- Django Model Utils version: master
- Django version: 2.1
- Python version: 3.6
- Other libraries used, if any:

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -3,6 +3,7 @@ CHANGES
 
 master (unreleased)
 -------------------
+- Catch `AttributeError` for deferred abstract fields, fixes GH-331.
 - Update documentation to explain usage of `timeframed` model manager, fixes GH-118
 - Honor `OneToOneField.parent_link=False`.
 - Fix handling of deferred attributes on Django 1.10+, fixes GH-278
@@ -14,7 +15,7 @@ master (unreleased)
 - Support `reversed` for all kinds of `Choices` objects, fixes GH-309
 - Fix Model instance non picklable GH-330
 - Fix patched `save` in FieldTracker
-- Upgrades test requirements (pytest, pytest-django, pytest-cov) and 
+- Upgrades test requirements (pytest, pytest-django, pytest-cov) and
   skips tox test with Python 3.5 and Django (trunk)
 
 3.1.2 (2018.05.09)

--- a/model_utils/tracker.py
+++ b/model_utils/tracker.py
@@ -40,7 +40,10 @@ class DescriptorWrapper(object):
         if instance is None:
             return self
         was_deferred = self.field_name in instance.get_deferred_fields()
-        value = self.descriptor.__get__(instance, owner)
+        try:
+            value = self.descriptor.__get__(instance, owner)
+        except AttributeError:
+            value = self.descriptor
         if was_deferred:
             tracker_instance = getattr(instance, self.tracker_attname)
             tracker_instance.saved_data[self.field_name] = deepcopy(value)

--- a/tests/models.py
+++ b/tests/models.py
@@ -224,6 +224,13 @@ class FeaturedManager(models.Manager):
         return ByAuthorQuerySet(self.model, **kwargs).filter(feature=True)
 
 
+class AbstractTracked(models.Model):
+    number = 1
+
+    class Meta:
+        abstract = True
+
+
 class Tracked(models.Model):
     name = models.CharField(max_length=20)
     number = models.IntegerField()
@@ -238,6 +245,14 @@ class TrackedFK(models.Model):
     tracker = FieldTracker()
     custom_tracker = FieldTracker(fields=['fk_id'])
     custom_tracker_without_id = FieldTracker(fields=['fk'])
+
+
+class TrackedAbstract(AbstractTracked):
+    name = models.CharField(max_length=20)
+    number = models.IntegerField()
+    mutable = MutableField(default=None)
+
+    tracker = FieldTracker()
 
 
 class TrackedNotDefault(models.Model):

--- a/tests/test_fields/test_field_tracker.py
+++ b/tests/test_fields/test_field_tracker.py
@@ -8,7 +8,7 @@ from model_utils import FieldTracker
 from model_utils.tracker import DescriptorWrapper
 from tests.models import (
     Tracked, TrackedFK, InheritedTrackedFK, TrackedNotDefault, TrackedNonFieldAttr, TrackedMultiple,
-    InheritedTracked, TrackedFileField,
+    InheritedTracked, TrackedFileField, TrackedAbstract,
     ModelTracked, ModelTrackedFK, ModelTrackedNotDefault, ModelTrackedMultiple, InheritedModelTracked,
 )
 
@@ -785,3 +785,8 @@ class InheritedModelTrackerTests(ModelTrackerTests):
         self.name2 = 'test'
         self.assertEqual(self.tracker.previous('name2'), None)
         self.assertTrue(self.tracker.has_changed('name2'))
+
+
+class AbstractModelTrackerTests(FieldTrackerTestCase):
+
+    tracked_class = TrackedAbstract

